rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,427 @@
+
+**Milestone 3: Optimized & generalized training (e.g. control flow, RNN)**<br/>
+Goal: Be able to train most models that are trainable with TensorFlow. Optimize the training performance for commonly used architectures.
+
+The following diagram gives an overview of TensorFlow Lite training road map.
+Yellow blocks are deliverables. Blue blocks are technical tasks to unblock the
+deliverables. The details of technical tasks will be explained below.
+
+<img src="20190722-tflite-training/roadmap.png" />
+
+## Goals & Non-Goals
+
+Goals:
+
+*   Illustrate a complete roadmap for TensorFlow Lite on-device training
+*   Describe user experience of TensorFlow Lite on-device training
+*   Design for functional on-device training of TensorFlow 2.x models with
+    TensorFlow Lite
+
+Non-goals:
+
+*   Quantized training on device
+*   Support TensorFlow Lite training with legacy model or TensorFlow 1.x
+
+## User Experience
+
+Throughout this document, **users** means developers who use TensorFlow Lite.
+
+This section explains how users can author a TensorFlow graph, convert a
+TensorFlow graph to TensorFlow Lite format, use TensorFlow Lite to run training
+and inference, and test TensorFlow Lite model correctness with the proposed
+design.
+
+Note: The example defines a single TensorFlow Lite model with multiple subgraphs
+for training & inference. However this isn't the only way -- You can also define
+separated training model & inference model.
+
+We're using a simplified example in this section:
+
+*   A model with 2 convolutional layers, 2 dense layers followed by a softmax
+    activation
+*   The whole model is trained offline
+*   The user want to retrain the 2 dense layers using personalized data on
+    device
+
+### Authoring TensorFlow graph
+
+**Defining the model using Keras Layers**
+
+The code defines the entire model using Keras API. Since we want to be retrain
+only dense layers on device, the model is defined in 2 parts (`conv_layers` and
+`dens_layer`).
+
+```python
+conv_layers = tf.keras.Sequential([
+  tf.keras.layers.Conv2D(10, kernel_size=(3, 3), activation="relu"),
+  tf.keras.layers.Conv2D(3, kernel_size=(3, 3), activation="relu"),
+  tf.keras.layers.Flatten(),
+])
+
+dense_layers = tf.keras.Sequential([
+  tf.keras.layers.Dense(50),
+  tf.keras.layers.Dense(10, activation='softmax'),
+])
+
+model = tf.keras.Sequential([conv_layers, dense_layers])
+```
+
+**Creating TensorFlow Functions for training and inference**
+
+With TensorFlow 2.0, the recommended way to use TensorFlow Lite is: Define a
+TensorFlow function for each invocable behavior in TensorFlow Lite. The
+converter will convert each TensorFlow function to a TensorFlow Lite subgraph,
+and users can choose subgraphs to invoke.
+
+The following code defines `inference` and `train` TensorFlow functions.
+
+```python
+@tf.function(input_signature=[
+    tf.TensorSpec(shape=[None, 64, 64, 3], dtype=tf.float32)])
+def inference(x):
+  return model(x)
```

Comment:
In TF 2.0, the TensorFlow Lite converter takes a `ConcreteFunction` as an input, so wrapping to-be-converted things inside `tf.function` becomes a canonical way. See also [this doc](https://www.tensorflow.org/lite/r2/convert/concrete_function). 

LMK if I understand your question correctly. 

Related your other comment, indeed I think it will be better if we can provide an easier to use interface for Keras.